source/egg/core/BitFlag.hh

```diff
@@ -0,0 +1,259 @@
+    /// @return True if all of the specified bits are off, otherwise false.
+    template <typename... Es>
+        requires(std::is_same_v<Es, E> && ...)
+    constexpr bool offBit(Es... es) const {
+        return (offBit_(es) && ...);
+    }
+
+    /// @brief Creates an applied mask of the corresponding bits for the provided enum values.
+    /// @details Matches the expression `(bits & mask)`.
+    /// @tparam ...Es Variadic template for packing.
+    /// @param ...es Enum values representing the bits to check.
+    /// @return The applied mask for the specified bits.
+    template <typename... Es>
+        requires(std::is_same_v<Es, E> && ...)
+    constexpr T maskBit(Es... es) const {
+        return bits & makeMask(es...);
+    }
+
+    /// @brief Creates a mask of the corresponding bits for the provided enum values.
+    /// @tparam ...Es Variadic template for packing.
+    /// @param ...es Enum values representing the bits to check.
+    /// @return The mask for the specified bits.
+    template <typename... Es>
+        requires(std::is_same_v<Es, E> && ...)
+    constexpr T makeMask(Es... es) const {
+        return (makeMask_(es) | ...);
+    }
+
+    /// @brief Sets the bits using a direct mask.
+    /// @param mask The bit mask to set.
+    /// @return Reference for chaining.
+    constexpr TBitFlag<T, E> &set(T mask) {
+        bits |= mask;
+        return *this;
+    }
+
+    /// @brief Resets the bits using a direct mask.
+    /// @param mask The bit mask to reset.
+    /// @return Reference for chaining.
+    constexpr TBitFlag<T, E> &reset(T mask) {
+        bits &= ~mask;
+        return *this;
+    }
+
+    /// @brief Changes the bits using a direct mask.
+    /// @param on Determines whether to set or reset the bits.
+    /// @param mask The bit mask to change.
+    /// @return Reference for chaining.
+    constexpr TBitFlag<T, E> &change(bool on, T mask) {
+        return on ? set(mask) : reset(mask);
+    }
+
+    /// @brief Checks if any bits are on in the specified mask.
+    /// @param mask The bit mask to check.
+    /// @return True if any bits in the mask are on, otherwise false.
+    constexpr bool on(T mask) const {
+        return (bits & mask) != 0;
+    }
+
+    /// @brief Checks if all bits are off in the specified mask.
+    /// @param mask The bit mask to check.
+    /// @return True if all bits in the mask are off, otherwise false.
+    constexpr bool off(T mask) const {
+        return (bits & mask) == 0;
+    }
+
+    /// @brief Resets all the bits to zero.
+    constexpr void makeAllZero() {
+        bits = 0;
+    }
+
+    /// @brief Gets the current bit mask.
+    /// @return The current bit mask.
+    constexpr T getDirect() const {
+        return bits;
+    }
+
+    /// @brief Sets the bits using a direct mask.
+    /// @param mask The bit mask to set.
+    constexpr void setDirect(T mask) {
+        bits = mask;
+    }
+
+private:
+    typedef std::underlying_type_t<E> EI;
```

Comment:
But T is the underlying integral type. Why are we using this instead?

Reply:
Nevermind, I see that T is the `TBitFlag` underlying type, but not necessarily `E`'s underlying type.